src/lib/grammar.rs

```diff
@@ -1,166 +1,84 @@
-use std::collections::{HashMap, HashSet};
-use std::fmt;
+use std::collections::HashMap;
 
-pub struct Grammar {
-    pub start: String,
-    pub rules: HashMap<String, Rule>,
-    pub tokens: HashSet<String>
-}
+use grammar_ast;
+
+/// A type specifically for nonterminal indices (i.e. corresponding to a rule
+/// name).
+pub type NIdx = i32;
```

Comment:
Any particular reason for using `i32` here? All `len()` methods for instance always return `usize`. I have to admit I'm not sure about their differences.